### DISCUSSION TITLE
How to expose ZIO executor metrics?

### DISCUSSION THREAD

**Discussion by jamesbassett** [zio/zio#9257] [Q&A]
I'm trying to expose the metrics of the ZIO executor (i.e. [these ones](https://github.com/zio/zio/blob/series/2.x/core/shared/src/main/scala/zio/internal/ExecutionMetrics.scala)) via prometheus so we can better understand how ZIO is performing.

I can't see any reference to these anywhere, and even searching for code references can only see them being used by the now deprecated [ZMX project](https://github.com/zio/zio-zmx/blob/b3226add36aee4f49d5644e862866735ae9df13c/core/jvm/src/main/scala/zio/zmx/diagnostics/protocol/Command.scala#L30).

I realise they're internal, but are these intended to be used at all? I was expecting them to be exposed with `OpenTelemetry.zioMetrics` but they are not.

Otherwise I'll likely have to use Micrometer's `ExecutorServiceMetrics.monitor`

**Comment by ghostdogpr** [comment]
They are not exposed as ZIO metrics, but you can access them by calling `ZIO.executor` and then the `metrics` method.

There are other internal metrics that are exposed: https://github.com/zio/zio/blob/ead44eb4bc650f6df454e0fc0581f2afc0ec0e37/core/shared/src/main/scala/zio/metrics/Metric.scala#L387

**Reply by jamesbassett** [reply]
Thanks, I'm using those other internal ones (they're exposed with `OpenTelemetry.zioMetrics` from zio opentelemetry).

Is there any plans to expose these executor metrics at all?  

For now I'm using Micrometer's `ExecutorServiceMetrics.monitor` to expose the thread pool metrics for the main and blocking executors, which is roughly the same info.

**Reply by ghostdogpr** [reply]
There is no particular plan AFAIK. I think if someone did a PR to add them, it would be okay as long as it doesn't impact performance.
